action/protocol/staking/protocol.go

```diff
@@ -684,3 +684,37 @@ func writeCandCenterStateToStateDB(sm protocol.StateManager, name, op, owners Ca
 	_, err := sm.PutState(owners, protocol.NamespaceOption(CandsMapNS), protocol.KeyOption(_ownerKey))
 	return err
 }
+
+// isSelfStakeBucket returns true if the bucket is self-stake bucket and not expired
+func isSelfStakeBucket(featureCtx protocol.FeatureCtx, csm CandidateStateManager, index uint64) (bool, error) {
+	// bucket index should be settled in one of candidates
+	selfStake := csm.ContainsSelfStakingBucket(index)
+	if featureCtx.DisableDelegateEndorsement || !selfStake {
+		return selfStake, nil
+	}
+
+	bucket, err := csm.getBucket(index)
+	switch {
+	case err == nil:
+	case errors.Is(err, ErrWithdrawnBucket):
```

Comment:
this case should not hit? if the bucket is withdrawn, L691 will = false?

Reply:
will hit, b/c unstake selfstake bucket will not clear `selfStakeBucketId`